mailpoet/lib/Segments/SubscribersFinder.php

```diff
@@ -46,35 +45,43 @@ private function findSubscribersInSegment(SegmentEntity $segment, $subscribersTo
     }
   }
 
-  private function isStaticSegment(Segment $segment) {
-    return in_array($segment->type, [Segment::TYPE_DEFAULT, Segment::TYPE_WP_USERS, Segment::TYPE_WC_USERS], true);
-  }
-
-  public function addSubscribersToTaskFromSegments(ScheduledTask $task, array $segments) {
+  /**
+   * @param ScheduledTask $task
+   * @param array<int>    $segmentIds
+   *
+   * @return float|int
+   */
+  public function addSubscribersToTaskFromSegments(ScheduledTask $task, array $segmentIds) {
     // Prepare subscribers on the DB side for performance reasons
-    $staticSegments = [];
-    $dynamicSegments = [];
-    foreach ($segments as $segment) {
-      if ($this->isStaticSegment($segment)) {
-        $staticSegments[] = $segment;
-      } elseif ($segment->type === SegmentEntity::TYPE_DYNAMIC) {
-        $dynamicSegments[] = $segment;
+    $staticSegmentIds = [];
+    $dynamicSegmentIds = [];
+    foreach ($segmentIds as $segment) {
+      $segment = $this->segmentsRepository->findOneById($segment);
+      if ($segment instanceof SegmentEntity) {
```

Comment:
Just a detail, but we could use early return here to save us from adding one level of nesting.
`if (!$segment instanceof SegmentEntity) continue;`

Reply:
That is a valid point. To be honest I'm not sure what version I prefer. I agree that having one less level of nesting helps with the readability of the code. On the other hand, I feel like "if nots" like `if (!$segment instanceof SegmentEntity)` are harder to grasp than "affirmative ifs" like `if ($segment instanceof SegmentEntity)`. Since I'm unsure I'm opting to leave the code as is. Happy to hear more about this if you feel like continuing this conversation.

Reply:
In general, I prefer the early termination approach because it allows adding a couple separated checks. After them, I'm safe to proceed with the main task. In my opinion, this approach works nicely in a function body. I agree that negative conditions might be harder to read. Especially `instanceof` as the `!` is placed in a strange place. It always felt to me like a negation of the variable on the left side. So a good point about the affirmative if in this case! I'm maybe just more used to the early termination approach.